### PR TITLE
OboeTester: add menu to hang and glitch

### LINKED ISSUE #1829: OboeTester: add ability to hang stream for N msec
This will help us debug
[b/286198906](http://b/286198906) | P2 | AAudio cannot recover from audio underrun if audio underrun time > 1000ms

Maybe add a menu to cause a periodic hang that lasts for 0,10,30,100,300,1000,3000 msec.

This could be in a separate Activity. TEST OUTPUT is getting crowded.

### PR DESCRIPTION
**Pull request by philburk**
Useful for testing recovery from hangs in a callback. Look for "Hang" menu under RELEASE button.

Fixes #1829

### CODE REVIEW

File: apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp [hunk 1]
```diff
@@ -38,3 +42,41 @@ void OboeTesterStreamCallback::printScheduler() {
...
+    if (startNanos > mNextTimeToHang) {
+        LOGD("%s() start sleeping", __func__);
+        // Take short naps until it is time to wake up.
+        int64_t nowNanos = startNanos;
+        int64_t wakeupNanos = startNanos + (mHangTimeMillis * 1000L * 1000L);
+        while (nowNanos < wakeupNanos && mHangTimeMillis > 0) {
+            int32_t sleepTimeMicros = (int32_t) ((wakeupNanos - nowNanos) / 1000);
+            // Don't usleep for more than a second.
+            sleepTimeMicros = std::min(100 * 1000, sleepTimeMicros);
+            if (sleepTimeMicros == 0) break;
```
Comment (flamme) [inline apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp:59]: Will there be a chance that `sleepTimeMicros` be zero? Won't it skip the while loop from the `nowNanos < wakeupNanos`?

File: apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp [hunk 2]
```diff
@@ -38,3 +42,41 @@ void OboeTesterStreamCallback::printScheduler() {
...
+
+    if (startNanos > mNextTimeToHang) {
+        LOGD("%s() start sleeping", __func__);
+        // Take short naps until it is time to wake up.
+        int64_t nowNanos = startNanos;
+        int64_t wakeupNanos = startNanos + (mHangTimeMillis * 1000L * 1000L);
+        while (nowNanos < wakeupNanos && mHangTimeMillis > 0) {
+            int32_t sleepTimeMicros = (int32_t) ((wakeupNanos - nowNanos) / 1000);
+            // Don't usleep for more than a second.
+            sleepTimeMicros = std::min(100 * 1000, sleepTimeMicros);
```
Comment (flamme) [inline apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp:58]: From the following line, isn't it 100ms?

File: apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp [hunk 3]
```diff
@@ -38,3 +42,41 @@ void OboeTesterStreamCallback::printScheduler() {
...
+
+// Sleep to cause an XRun. Then reschedule.
+void OboeTesterStreamCallback::maybeHang(const int64_t startNanos) {
+    if (mHangTimeMillis == 0) return;
+
+    if (startNanos > mNextTimeToHang) {
+        LOGD("%s() start sleeping", __func__);
+        // Take short naps until it is time to wake up.
+        int64_t nowNanos = startNanos;
+        int64_t wakeupNanos = startNanos + (mHangTimeMillis * 1000L * 1000L);
```
Comment (flamme) [inline apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp:54]: Would you mind making a constant here given it is also used at line 68?

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestAudioActivity.java [hunk 4]
```diff
@@ -437,6 +443,24 @@ public void onClick(View v) {
         }
         OboeAudioStream.setCallbackReturnStop(false);
 
+        mHangTimeSpinner = (Spinner) findViewById(R.id.spinner_hang_time);
```
Comment (robertwu1) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestAudioActivity.java:446]: What happens for activities for that open both input and output streams? Do both get hung by the same amount

File: apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp [hunk 5]
```diff
@@ -38,3 +42,41 @@ void OboeTesterStreamCallback::printScheduler() {
...
+        while (nowNanos < wakeupNanos && mHangTimeMillis > 0) {
+            int32_t sleepTimeMicros = (int32_t) ((wakeupNanos - nowNanos) / 1000);
+            // Don't usleep for more than a second.
```
Comment (robertwu1) [inline apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp:57]: I'm a bit confused by these comments given that the min setting is 4ms and the max is 5 seconds

Comment (philburk) [reply apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp:57]: I improved the comments.

File: apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp (hunk 2 above)
Comment (philburk) [reply apps/OboeTester/app/src/main/cpp/OboeTesterStreamCallback.cpp:58]: Yes, this is 100 ms.

**Review by robertwu1 — APPROVED** [review]

**Review by flamme — APPROVED** [review]

**Pull request merged**